File: c++/metrics.cpp
#include "matrix.h"
#include "metrics.h"
#include <iostream>
namespace metric
{
    /* accuracy()
	* -----
	* Computes the Accuracy metric, that is: 
	* Accuracy = Number of correct predictions/total number of predictions.
	*
	* @y_true: the true labels
	* @y_pred: the predicted labels, i.e. output from the last layer, 
	*
	* Returns: the calculated accuracy
	*/
    double accuracy(matrix::Matrix<double> y_true, 
			matrix::Matrix<double> y_pred) {

		matrix::Matrix<int> prediction = matrix::argmax(y_pred);
		matrix::Matrix<int> tru = matrix::argmax(y_true);
		return (matrix::sum(matrix::equals(tru, prediction), 0, false)[0][0] /  
				(double) y_pred.rows);
    }
}

File: python/Metrics.py
from abc import abstractmethod
import numpy as np
from abc import ABC

class Metric(ABC):

    @abstractmethod
    def calculate():
        pass

class Accuracy(Metric):

    def calculate(self, y_true, y_pred):
        prediction = np.argmax(y_pred, axis=1)
        # If y_true are one-hot
        if len(y_true.shape) > 1:
            y_true = np.argmax(y_true, axis=1)
        return np.mean(y_true == prediction)